### ISSUE TITLE
Yielded values are not recognised as returned values in Squiz FunctionComment sniff

### ISSUE THREAD

**Issue by haeber** [squizlabs/PHP_CodeSniffer#479]
Currently there is no type hint for a return value built with yield like the following 1st example.
```PHP
/**
 * @return array
 */
function yieldTest()
{
    for ($i = 0; $i < 5; $i++) {
        yield $i;
    }
}

foreach (yieldTest() as $i) {
    // ...
} 
```

Currently CodeSniffer does not understand the @yield annotation, furthermore it does not interpret the @return annotation with yield correctly. There is the first case like in this example (with function yieldTest) where CodeSniffer complains about the missing return statement
> Function return type is not void, but function has no return statement

There is another case where code sniffer complains about the following 2nd example (with usage of function yieldTest2), when @return void is used.
```PHP
/**
 * @return void
 */
function yieldTest2()
{
    for ($i = 0; $i < 5; $i++) {
        yield $i;
    }
}

foreach (yieldTest2() as $i) {
    // ...
} 
```

The message is:
> 'void' method yieldTest2 result used

So please support one case for the usage of yield. Weather with @return or @yield, so there is one right way for well formed code without a false positive warning.

Kind Regards
Thomas

**Comment by aik099** [comment]
> Function return type is not void, but function has no return statement

Are you sure, that it's PHP_CodeSniffer that is checking that, not IDE (e.g. PhpStorm) inspection?

PHP_CodeSniffer doesn't look for `return` statements in the function to determine if `@return` tag should be present in function's DocBlock.

**Comment by Konafets** [comment]
Actually PHPCS does do that.

https://github.com/squizlabs/PHP_CodeSniffer/blob/master/CodeSniffer/Standards/Squiz/Sniffs/Commenting/FunctionCommentSniff.php#L46

**Comment by haeber** [comment]
@aik099 The message
> Function return type is not void, but function has no return statement
comes from PHPCS, but I think that the 2nd message from my 2nd example comes from PHPStorm.

@Konafets What does this implementation mean, if it still produces the 1st message, shall I use @return void?

**Comment by Konafets** [comment]
@haeber My comment was for @aik099 as he said, that PHPCS not checking for return in functions. Beside this I can't say anything in regard your issue, I heard the first time about yield just right now.

**Comment by aik099** [comment]
Technically `yield` can be considered as return in the sniff code, but I'm not sure if that case is mentioned anywhere on http://www.phpdoc.org/ .

**Comment by gsherwood** [comment]
That sniff wont complain any more. Yeilded values are seen as normal return values.

**Comment by haeber** [comment]
Thanks a lot.
